### PR TITLE
feat(core): add LweCiphertextDiscardingCircuitBootstrapBooleanEngine

### PR DESCRIPTION
**Pull request by IceTDrinker**
### Resolves:

closes https://github.com/zama-ai/concrete-core-internal/issues/397
closes https://github.com/zama-ai/concrete-core-internal/issues/398
closes https://github.com/zama-ai/concrete-core-internal/issues/399

### Description

- add spec and impls for 32/64 bits

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (comment with @slab-ci cpu_test and/or @slab-ci gpu_test to trigger the tests)
* [x] The draft release description has been updated
* [x] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+impl<EngineError: std::error::Error>
+    LweCiphertextDiscardingCircuitBootstrapBinaryError<EngineError>
+{
+    /// Validates the inputs
+    pub fn perform_generic_checks() -> Result<(), Self> {
+        Ok(())
+    }
+}
+
+/// A trait for engines performing a (discarding) circuit bootstrap on an LWE ciphertext containing
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs:23]: containing -> encrypting?

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+engine_error! {
+    LweCiphertextDiscardingCircuitBootstrapBinaryError for
+    LweCiphertextDiscardingCircuitBootstrapBinaryEngine @
+}
+
+impl<EngineError: std::error::Error>
+    LweCiphertextDiscardingCircuitBootstrapBinaryError<EngineError>
+{
+    /// Validates the inputs
+    pub fn perform_generic_checks() -> Result<(), Self> {
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs:18]: Here the generic checks should be added I guess.

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+/// A trait for engines performing a (discarding) circuit bootstrap on an LWE ciphertext containing
+/// a binary message (i.e. containing only 1 bit of information).
+///
+/// # Semantics
+///
+/// This [discarding](super#operation-semantics) operation fills the `output` GGSW ciphertext with
+/// the result of the circuit bootstrap of the `input` LWE ciphertext using the given bootstrap key
+/// `bsk` and vector of private functional packing keyswitch keys `cbs_pfpksk`.
+///
+/// # Formal Definition
```
Comment (agnesLeroy) [inline concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs:32]: We should open an issue for all the woPBS formal definitions.

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    /// ```
+    fn discard_circuit_bootstrap_binary_lwe_ciphertext(
+        &mut self,
+        output: &mut GgswCiphertext32,
+        input: &LweCiphertext32,
+        delta_log: DeltaLog,
+        bsk: &FftwFourierLweBootstrapKey32,
+        cbs_pfpksk: &LweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys32,
+    ) -> Result<(), LweCiphertextDiscardingCircuitBootstrapBinaryError<Self::EngineError>> {
+        LweCiphertextDiscardingCircuitBootstrapBinaryError::perform_generic_checks()?;
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs:130]: Here we should check allowed poly size as well.

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+    /// let cbs_pfpksk: LweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys32 = default_engine
+    ///     .generate_new_lwe_circuit_bootstrap_private_functional_packing_keyswitch_keys(
+    ///         &lwe_sk,
+    ///         &glwe_sk,
+    ///         base_log_pfpksk,
+    ///         level_pfpksk,
+    ///         noise,
+    ///     )?;
+    ///
+    /// // delta_log ndicates where the information bit is stored in the input LWE ciphertext, here
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs:89]: indicates

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+    /// let cbs_pfpksk: LweCircuitBootstrapPrivateFunctionalPackingKeyswitchKeys64 = default_engine
+    ///     .generate_new_lwe_circuit_bootstrap_private_functional_packing_keyswitch_keys(
+    ///         &lwe_sk,
+    ///         &glwe_sk,
+    ///         base_log_pfpksk,
+    ///         level_pfpksk,
+    ///         noise,
+    ///     )?;
+    ///
+    /// // delta_log ndicates where the information bit is stored in the input LWE ciphertext, here
```
Comment (agnesLeroy) [inline concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs:227]: indicates

**Review by agnesLeroy — COMMENTED** [review]
Hello @IceTDrinker! Thanks a lot for this contribution, here comes my review :slightly_smiling_face:

File: concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs (hunk 4 above)
Comment (IceTDrinker) [reply concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs:130]: Yeah… lately it seems I’m always forgetting about those

Comment (agnesLeroy) [reply concrete-core/src/backends/fftw/implementation/engines/fftw_engine/lwe_ciphertext_discarding_circuit_bootstrap_binary.rs:130]: Can't blame you, remember my last PR about vector views! :wink:

File: concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs (hunk 3 above)
Comment (IceTDrinker) [reply concrete-core/src/specification/engines/lwe_ciphertext_vector_discarding_circuit_bootstrap_binary.rs:84]: https://github.com/zama-ai/concrete-core-internal/issues/401

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Comment by IceTDrinker** [comment]
updated mainly for the diff view for ongoing work

**Comment by agnesLeroy** [comment]
@IceTDrinker can we close this one?

**Comment by IceTDrinker** [comment]
No, I'll take some time to update it and follow the boolean notation, it can be useful to convert an LWE ciphertext to a GGSW representation

**Comment by IceTDrinker** [comment]
unless we say it's not necessary anymore

**Comment by IceTDrinker** [comment]
@slab-ci cpu_test

**Review by agnesLeroy — APPROVED** [review]

**Pull request merged**